docs/architecture/adr-024-coin-metadata.md

```diff
@@ -0,0 +1,104 @@
+regardless of the denomination itself. In other words, clients and apps built atop a Cosmos-SDK-based
+chain may choose to define and use arbitrary units of denomination to provide a richer UX, however, by
+the time a tx or operation reaches the Cosmos SDK state machine, the `amount` is treated as a single
+unit. For example, for the Cosmos Hub (Gaia), clients assume 1 ATOM = 10^6 uatom, and so all txs and
+operations in the Cosmos SDK work off of units of 10^6.
+
+This clearly provides a poor and limited UX especially as interoperability of networks increases and
+as a result the total amount of asset types increases. We propose to have `x/bank` additionally keep
+track of metadata per `denom` in order to help clients, wallet providers, and explorers improve their
+UX and remove the requirement for making any assumptions on the unit of denomination.
+
+## Decision
+
+The `x/bank` module will updated to store and index metadata by `denom`, specifically the "base" or
+smallest unit -- the unit the Cosmos SDK state-machine works with. The metadata should ideally be
+extensible and serve the needs of many external consumers and even other modules (e.g. `x/ibc` or `x/nft`).
+At a minimum, the metadata type should include all units and a potential list of denomination aliases.
+As a result, we can define the type as follows:
+
+```protobuf
+message DenomUnit {
+  string denom    = 1;
+  uint32 exponent = 2;  
+}
+
+message Metadata {
+  string description = 1;
+  string base = 2;
+  repeated DenomUnit denom_units = 3;
+  repeated string aliases = 4;
+}
+```
+
+As an example, the ATOM's metadata can be defined as follows:
+
+```json
+{
+  "description": "The native staking token of the Cosmos Hub.",
+  "base": "uatom",
+  "denom_units": [
+    {
+      "denom": "matom",
+      "exponent": 3,
+    },
+    {
+      "denom": "atom",
+      "exponent": 6,
+    }
+  ],
+  "aliases": [
+    "btcuatom",
+    "kavauatom"
+  ]
+}
+```
+
+> Given the above metadata, a client may infer that 4.3atom = 4.3 * (10^6) = 4,300,000uatom.
+
+A client should be able to query for metadata by denom both via the CLI and REST interfaces. In
+addition, we will add handlers to these interfaces to convert from any unit to another given unit,
+as the base framework for this already exists in the Cosmos SDK.
+
+## Future Work
+
+In order for clients to avoid having to convert assets to the base denomination -- either manually or
+via an endpoint, we may consider supporting automatic conversion of a given unit input.
+
+## Consequences
+
+### Positive
+
+- Provides clients, wallet providers and block explorers with additional data on
+  asset denomination to improve UX and remove any need to make assumptions on
+  denomination units.
+
+### Negative
+
+- A small amount of required additional storage in the `x/bank` module. The amount
+  of additional storage should be minimal as the amount of total assets should not
+  be large.
```

Comment:
> The amount of additional storage should be minimal as the amount of total assets should not be large.

Well depends on what are we using as the store key. Eg: if a coin is relayed thought different connections on IBC 

`A->X->C`, `A->Y->C`, `A->Z->C` will all have the different denominations for the same token. cc: @cwgoes

Reply:
maybe we can consider using the alias as the key: `<source-chain>|<original-denom>`

Reply:
Again, nothing to do with IBC or relaying. The ADR mentions the metadata is stored and indexed by the base denom and will contain a few fields, so it should be a few bytes at max.

Reply:
> The ADR mentions the metadata is stored and indexed by the base denom

On the IBC you can see the coin denominations like `transfer/ibczeroxfer/uatom` (in the format `{portIDchainA}/{channelIDchainA}/.../{portIDchainN}/{channelIDchainN}/denom`. So you will need to strip the denom and store use it as the key.

Reply:
That's not the base denom though and thus it's not what it's indexed by. In this case, the base denom is `uatom` and metadata for `uatom` exists on the chain in which it's native.

Reply:
> > The amount of additional storage should be minimal as the amount of total assets should not be large.
> 
> Well depends on what are we using as the store key. Eg: if a coin is relayed thought different connections on IBC
> 
> `A->X->C`, `A->Y->C`, `A->Z->C` will all have different denominations for the same underlying token. cc: @cwgoes

They aren't really the "same underlying token" - they may be redeemable for the same token if the reverse redemption path is taken, but Byzantine behaviour could always occur (thus we separate the denominations) - there is some extra storage cost but it is not that large & unavoidable.

I don't think we can use the alias as the key unless it is guaranteed to be globally unique, which is a stricter requirement than this ADR imposes (more appropriate to a formal "alias registry" somewhere).